plugins/pkg/file/fs.go

```diff
@@ -29,110 +28,75 @@ var (
 )
 
 type File struct {
-	lock sync.RWMutex
-
-	Name  string
-	mtime time.Time
-}
-
-func (f *File) Mtime() time.Time {
-	f.lock.RLock()
-	defer f.lock.RUnlock()
-	// the returned time.Time should be readonly
-	return f.mtime
-}
-
-func (f *File) SetMtime(t time.Time) {
-	f.lock.Lock()
-	f.mtime = t
-	f.lock.Unlock()
+	Name string
 }
 
-type fs struct {
-	cache *ttlcache.Cache[string, os.FileInfo]
+type Watcher struct {
+	watcher *fsnotify.Watcher
+	files   map[string]bool
+	mu      sync.Mutex
+	dir     map[string]bool
+	done    chan struct{}
 }
 
-func newFS(ttl time.Duration) *fs {
-	loader := ttlcache.LoaderFunc[string, os.FileInfo](
-		func(c *ttlcache.Cache[string, os.FileInfo], key string) *ttlcache.Item[string, os.FileInfo] {
-			info, err := os.Stat(key)
-			if err != nil {
-				logger.Error(err, "reload file info to cache", "file", key)
-				return nil
-			}
-			item := c.Set(key, info, ttlcache.DefaultTTL)
-			logger.Info("update file mtime", "file", key, "mtime", item.Value().ModTime())
-			return item
-		},
-	)
-	cache := ttlcache.New(
-		ttlcache.WithTTL[string, os.FileInfo](ttl),
-		ttlcache.WithLoader[string, os.FileInfo](loader),
-	)
-	go cache.Start()
-
-	return &fs{
-		cache: cache,
+func NewWatcher() (*Watcher, error) {
+	w, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
 	}
+	return &Watcher{
+		watcher: w,
+		files:   make(map[string]bool),
+		done:    make(chan struct{}),
+		dir:     make(map[string]bool),
+	}, nil
 }
 
-var (
-	// TODO: rewrite it to use inotify
-	defaultFs = newFS(10 * time.Second)
-)
-
-func IsChanged(files ...*File) bool {
+func (w *Watcher) AddFiles(files ...*File) error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
 	for _, file := range files {
-		changed := defaultFs.isChanged(file)
-		if changed {
-			return true
+		if _, exists := w.files[file.Name]; !exists {
+			w.files[file.Name] = true
+		}
+		dir := filepath.Dir(file.Name)
+		if _, exists := w.dir[dir]; !exists {
+			if err := w.watcher.Add(dir); err != nil {
+				return err
+			}
+			w.dir[dir] = true
 		}
 	}
-	return false
-}
-
-func (f *fs) isChanged(file *File) bool {
-	item := f.cache.Get(file.Name)
-	if item == nil {
-		// As a protection, failed to fetch the real file means file not changed
-		return false
-	}
-
-	return file.Mtime().Before(item.Value().ModTime())
-}
-
-func (f *fs) Stat(path string) (*File, error) {
-	info, err := os.Stat(path)
-	if err != nil {
-		return nil, err
-	}
-	f.cache.Set(path, info, ttlcache.DefaultTTL)
-
-	return &File{
-		Name:  path,
-		mtime: info.ModTime(),
-	}, nil
+	return nil
 }
 
-func Stat(path string) (*File, error) {
-	return defaultFs.Stat(path)
+func (w *Watcher) Start(onChanged func()) {
+	go func() {
+		logger.Info("start watching files")
+		for {
+			select {
+			case event := <-w.watcher.Events:
+				if _, exists := w.files[event.Name]; exists {
```

Comment:
The test case doesn't work on Mac. I spent some time investigating it and found that the event.Name is `example1104322652` and the map is `map[./example1104322652:true]`. Let's convert the input file name to absolute path.